plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -352,27 +367,47 @@ applyHint ide nfp mhint =
     -- If we provide "applyRefactorings" with "Just (1,13)" then
     -- the "Redundant bracket" hint will never be executed
     -- because SrcSpan (1,20,??,??) doesn't contain position (1,13).
-    let fp = fromNormalizedFilePath nfp
-    (_, mbOldContent) <- liftIO $ runAction "hlint" ide $ getFileContents nfp
-    oldContent <- maybe (liftIO $ T.readFile fp) return mbOldContent
-    -- We need to save a file with last edited contents cause `apply-refact`
-    -- doesn't expose a function taking directly contents instead a file path.
-    -- Ideally we should try to expose that function upstream and remove this.
-    res <- liftIO $ withSystemTempFile (takeFileName fp) $ \temp h -> do
+#ifdef GHC_LIB
```

Comment:
Is anyone using the GHC_LIB path?

Reply:
Digital Asset were using it for DAML. But they use Ghcide not HLS, so I believe that for HLS its totally unused. I'd just delete it.

Reply:
`GHC_LIB` is used by hlint for ghc < 8.10 and we are following it closely. This is [the significative part of its .cabal file](https://github.com/mpickering/apply-refact/issues/95#issuecomment-731926466):

```cabal

    if !flag(ghc-lib) && impl(ghc >= 8.10.0) && impl(ghc < 8.11.0)
      build-depends:
        ghc == 8.10.*,
        ghc-boot-th,
        ghc-boot
    else
      build-depends:
          ghc-lib-parser == 8.10.*
```

As ghcide and apply-refact are using the "real" ghc for all versions we have some friction between them.

But your question make me think that maybe we dont have to use it for `apply-refact` and pass the parsed module in all cases. Only hlint requires ghc-lib.
Not sure if that dissymmetry could cause issues though. Moreover, there is an issue in apply-refact to make it use ghc-lib, logical move if we take in account both packages are closely related.

Reply:
Ah, that makes sense, its quite a different purpose GHC_LIB to the one in Ghcide.

Reply:
If this defined variable serves a different purpose, then maybe give it a different name to avoid confusion.

But I still don't understand what it means. Where/when is it defined?

Reply:
Well, afaiu for all cases it means you want to use ghc-lib, usually instead ghc:
* in ghcide the default is to use real ghc for all ghc versions and the cabal flag makes the package use ghc-lib for *all* ghc versions
  * the main goal is support daml
* in hlint the default is use ghc-lib for ghc < 8.10 and ghc for 8.10, the cabal flag makes hlint use ghc-lib for all versions
  * afaik the main goal is to avoid a ghc compat module, as ghc-lib is *the* ghc compat package

We copied in the plugin hlint cabal definitions:

https://github.com/haskell/haskell-language-server/blob/d0fb85eb2c35d0c5ad0e4d0610d89e05d37f5b26/plugins/hls-hlint-plugin/hls-hlint-plugin.cabal#L19-L23

https://github.com/haskell/haskell-language-server/blob/d0fb85eb2c35d0c5ad0e4d0610d89e05d37f5b26/plugins/hls-hlint-plugin/hls-hlint-plugin.cabal#L55-L64

But we have to use ghc and ghc-lib at the same time, to make a bridge between ghcide and hlint.

The `GHC-LIB` cpp condition guards code that have to take in account ghc-lib (like in ghcide, no?). Not sure what other better name could have.

Reply:
A suitable name would be HLINT_ON_GHC_LIB perhaps?

Reply:
Mmm, maybe a little bit too close to the actual implementation (i can imagine apply-refact also using ghc-lib f.e.) but it is clear in exchange, will do